pageserver/src/tenant.rs

```diff
@@ -131,6 +133,257 @@ pub enum TenantState {
+        &self,
+        reader: impl std::io::Read,
+        base_lsn: Lsn,
+    ) -> anyhow::Result<()> {
+        let raw_timeline = self.raw_timeline()?;
+        import_datadir::import_basebackup_from_tar(raw_timeline, reader, base_lsn).with_context(
+            || {
+                format!(
+                    "Failed to import basebackup for timeline {}/{}",
+                    self.owning_tenant.tenant_id, self.timeline_id
+                )
+            },
+        )?;
+
+        fail::fail_point!("before-checkpoint-new-timeline", |_| {
+            bail!("failpoint before-checkpoint-new-timeline");
+        });
+
+        raw_timeline
+            .checkpoint(CheckpointConfig::Flush)
+            .with_context(|| {
+                format!(
+                    "Failed to checkpoint after basebackup import for timeline {}/{}",
+                    self.owning_tenant.tenant_id, self.timeline_id
+                )
+            })?;
+        Ok(())
+    }
+
+    fn raw_timeline(&self) -> anyhow::Result<&Timeline> {
+        Ok(&self
+            .raw_timeline
+            .as_ref()
+            .with_context(|| {
+                format!(
+                    "No raw timeline {}/{} found",
+                    self.owning_tenant.tenant_id, self.timeline_id
+                )
+            })?
+            .0)
+    }
+}
+
+impl Drop for UninitializedTimeline<'_> {
+    fn drop(&mut self) {
+        if let Some((_, tombstone)) = self.raw_timeline.take() {
+            let tenant_id = self.owning_tenant.tenant_id;
+            let timeline_id = self.timeline_id;
+            let _entered = info_span!("drop_unitialized_timeline", tenant = %tenant_id, timeline = %timeline_id).entered();
+            error!("Timeline got dropped without initializing, cleaning its files");
+
+            let raw_timeline_dir = &tombstone.timeline_path;
+            let mut should_remove_tombstone = true;
+            match fs::remove_dir_all(raw_timeline_dir) {
+                Ok(()) => info!("Timeline dir removed successfully, removing the tombstone"),
+                Err(e) if e.kind() == io::ErrorKind::NotFound => {
+                    info!("Timeline dir is absent, skipping and removing the tombstone")
+                }
+                Err(e) => {
+                    should_remove_tombstone = false;
+                    error!("Failed to clean up unitialized timeline directory: {e:?}");
+                }
+            }
+
+            if should_remove_tombstone {
+                if let Err(e) = tombstone.remove_tombstone() {
+                    error!("Failed to remove the tombstone: {e:?}");
+                }
+            }
+        }
+    }
+}
+
+impl TimelineTombstone {
+    /// Useful for initializing timelines, existing on disk after the restart.
+    pub fn dummy() -> Self {
+        Self {
+            tombstone_deleted: true,
+            tombstone_path: PathBuf::new(),
+            timeline_path: PathBuf::new(),
+        }
+    }
```

Comment:
This feels a bit weird. I think changing the tombstone field in UninitializedTimeline into `Option<TimelineTombstone>` would be more natural, and use None when there is no tombstone file.

Reply:
Or perhaps have different structs for the different cases:
- `CreateInProgressTimeline` for a newly created branch, with tombstone to delete it on crash
- `LoadInProgressTimeline` for a timeline that exists on disk, and we're loading it. No tombstone file.

Reply:
>  I think changing the tombstone field in UninitializedTimeline into `Option<TimelineTombstone>` would be more natural, and use None when there is no tombstone file.

I've went with the simplest approach later naturally too, and now the tombstone looks like 
```rust
#[must_use]
pub struct UninitializedTimeline<'t> {
    owning_tenant: &'t Tenant,
    timeline_id: TimelineId,
    raw_timeline: Option<(Timeline, TimelineTombstone)>,
}
```
which, presumably matches the proposal.